src/Development/IDE/Spans/AtPoint.hs

```diff
@@ -23,35 +29,110 @@ import Development.IDE.Spans.Common
 import Development.IDE.Core.RuleTypes
 
 -- GHC API imports
-import FastString
 import Name
 import Outputable hiding ((<>))
 import SrcLoc
 import TyCoRep
 import TyCon
 import qualified Var
 import NameEnv
+import Module
+import IfaceType
 
 import Control.Applicative
 import Control.Monad.Extra
 import Control.Monad.Trans.Maybe
 import Control.Monad.Trans.Class
 import Control.Monad.IO.Class
 import           Data.Maybe
-import           Data.List
 import qualified Data.Text as T
 import qualified Data.Map as M
+import qualified Data.HashMap.Strict as HM
+import Development.IDE.Core.PositionMapping
 
 import Data.Either
-import Data.List.Extra (dropEnd1)
+import Data.List.Extra (nubOrd, dropEnd1)
+
+import HieDb hiding (pointCommand)
+import qualified Data.Array as A
+
+-- | Gives a Uri for the module, given the .hie file location and the the module info
+-- The Bool denotes if it is a boot module
+type LookupModule m = FilePath -> ModuleName -> UnitId -> Bool -> MaybeT m Uri
+
+-- | HieFileResult for files of interest, along with the position mappings
+newtype FOIReferences = FOIReferences (HM.HashMap NormalizedFilePath (HieAstResult, PositionMapping))
+
+-- | Given a file and position, return the names at a point, the references for
+-- those names in the FOIs, and a list of file paths we already searched through
+foiReferencesAtPoint
+  :: NormalizedFilePath
+  -> Position
+  -> FOIReferences
+  -> ([Name],[Location],[FilePath])
+foiReferencesAtPoint file pos (FOIReferences asts) =
+  case HM.lookup file asts of
+    Nothing -> ([],[],[])
+    Just (HAR _ hf _ _,mapping) ->
+      let posFile = fromMaybe pos $ fromCurrentPosition mapping pos
+          names = concat $ pointCommand hf posFile (rights . M.keys . nodeIdentifiers . nodeInfo)
+          adjustedLocs = HM.foldr go [] asts
+          go (HAR _ _ rf _, mapping) xs = refs ++ xs
+            where
+              refs = mapMaybe (toCurrentLocation mapping . realSrcSpanToLocation . fst)
+                   $ concat $ mapMaybe (\n -> M.lookup (Right n) rf) names
+          toCurrentLocation mapping (Location uri range) = Location uri <$> toCurrentRange mapping range
+        in (names, adjustedLocs,map fromNormalizedFilePath $ HM.keys asts)
+
+referencesAtPoint
+  :: MonadIO m
+  => HieDb
+  -> NormalizedFilePath
+  -> Position
+  -> FOIReferences
+  -> m [Location]
+referencesAtPoint hiedb nfp pos refs = do
+  let (names, foisLocs, exclude) = foiReferencesAtPoint nfp pos refs
+  locs <- forM names $ \name ->
+    case nameModule_maybe name of
+      Nothing -> pure []
+      Just mod -> do
+         rows <- liftIO $ search hiedb True (nameOccName name) (Just $ moduleName mod) (Just $ moduleUnitId mod) exclude
```

Comment:
What does `search` do?

Reply:
It returns all references from the database, except for the ones that occur in `exclude`.